lib/active_record/connection_adapters/trilogy_adapter.rb

```diff
@@ -153,13 +153,87 @@ def discard!
         self.connection = nil
       end
 
+      # ActiveRecord 7.0 support
+      if ActiveRecord.version < ::Gem::Version.new('7.1.a')
+        def initialize(connection, logger, connection_options, config)
+          super
+          if @connection
+            @verified = true
+            @raw_connection = @connection
+          end
+          # Ensure that we're treating prepared_statements in the same way that Rails 7.1 does
+          @prepared_statements = self.class.type_cast_config_to_boolean(
+            @config.fetch(:prepared_statements) { default_prepared_statements }
+          )
+        end
+
+        def connect!
+          verify!
+          self
+        end
+
+        def reconnect!
+          @lock.synchronize do
+            disconnect!
+            connect
+          rescue StandardError => original_exception
+            @verified = false
+            raise translate_exception_class(original_exception, nil, nil)
+          end
+        end
+
+        def exec_rollback_db_transaction
+          # 16384 tests the bit flag for SERVER_SESSION_STATE_CHANGED, which gets set when the
+          # last statement executed has caused a change in the server's state.
+          if active? || (@raw_connection.server_status & 16_384).positive?
```

Comment:
Q: what is the reason for checking the server status ? IMO rollback should be a no-op if the server doesn't have any modifications right ?

Reply:
After careful analysis, I've found that this check for session state changed becomes useful if we were to support AR 6.1 and older, so it could be removed -- will drop in a commit for this later this evening.

(I do appreciate your interest in being thorough!)